packages/opentelemetry-plugin-xml-http-request/src/xhr.ts

```diff
@@ -0,0 +1,370 @@
+      addSpanNetworkEvent(span, PTN.FETCH_START, resource);
+      addSpanNetworkEvent(span, PTN.DOMAIN_LOOKUP_START, resource);
+      addSpanNetworkEvent(span, PTN.DOMAIN_LOOKUP_END, resource);
+      addSpanNetworkEvent(span, PTN.CONNECT_START, resource);
+      addSpanNetworkEvent(span, PTN.SECURE_CONNECTION_START, resource);
+      addSpanNetworkEvent(span, PTN.CONNECT_END, resource);
+      addSpanNetworkEvent(span, PTN.REQUEST_START, resource);
+      addSpanNetworkEvent(span, PTN.RESPONSE_START, resource);
+      addSpanNetworkEvent(span, PTN.RESPONSE_END, resource);
+    }
+  }
+
+  /**
+   * Removes the previous information about span.
+   * This might happened when the same xhr is used again.
+   * @param xhr
+   * @private
+   */
+  private _cleanPreviousSpanInformation(xhr: XMLHttpRequestWrapped) {
+    const existingSpanId = xhr.__OT_SPAN_ID;
+    if (existingSpanId) {
+      const callbackToRemoveEvents = this._callbackToRemoveEvents[
+        existingSpanId
+      ];
+      if (callbackToRemoveEvents) {
+        callbackToRemoveEvents();
+      }
+    }
+  }
+
+  /**
+   * Creates a new span when method "open" is called
+   * @param xhr
+   * @param url
+   * @param method
+   * @private
+   */
+  private _createSpan(
+    xhr: XMLHttpRequestWrapped,
+    url: string,
+    method: string
+  ): tracing.Span | undefined {
+    if (isUrlIgnored(url, this._config.ignoreUrls)) {
+      this._logger.debug('ignoring span as url matches ignored url');
+      return;
+    }
+
+    const currentSpan = this._tracer.startSpan(url, {
+      parent: this._tracer.getCurrentSpan(),
+      attributes: {
+        [AttributeNames.COMPONENT]: this.component,
+        [AttributeNames.HTTP_METHOD]: method,
+        [AttributeNames.HTTP_URL]: url,
+      },
+    }) as tracing.Span;
+
+    currentSpan.addEvent(EventNames.METHOD_OPEN);
+
+    const spanId = currentSpan.context().spanId;
+    this._spans[spanId] = currentSpan;
+
+    this._cleanPreviousSpanInformation(xhr);
+    xhr.__OT_SPAN_ID = spanId;
+
+    return currentSpan;
+  }
+
+  /**
+   * Marks certain [resource]{@link PerformanceResourceTiming} when information
+   * from this is used to add events to span.
+   * This is done to avoid reusing the same resource again for next span
+   * @param resource
+   * @private
+   */
+  private _markResourceAsUsed(resource: PerformanceResourceTiming) {
+    const name = resource.name;
+    if (!this._usedResources[name]) {
+      this._usedResources[name] = [];
+    }
+    this._usedResources[name].push(resource);
+  }
+
+  /**
+   * Patches the method open
+   * @private
+   */
+  protected _patchOpen() {
+    return (original: OpenFunction): OpenFunction => {
+      const plugin = this;
+      return function patchOpen(
+        this: XMLHttpRequest,
+        method: string,
+        url: string,
+        async?: boolean,
+        user?: string | null,
+        pass?: string | null
+      ): OpenFunction | any {
```

Comment:
Incorrect return type. The `OpenFunction` is not returned, but its return value is.
```suggestion
      ): ReturnType<OpenFunction> {
```

Reply:
it should be `void` in fact